<a href="https://colab.research.google.com/github/PaulToronto/DeepLearning-AI---Mathematics-for-Machine-Learning-and-Data-Science/blob/main/3_1_Introduction_to_Probability_and_Probability_Distributions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 3.1 Introduction to Probability and Probability Distributions

## Imports

In [1]:
import pandas as pd

## 3.1.1 Introduction to Probability

### 3.1.1.1 What is Probability?

**Probability** is a measure of how likely and event is to occur.



#### Example

You are in a school with 10 kids and you randomly pick a kid. 3 of them play soccer and 7 of them don't. What is the probability that you will pick a kid who plays soccer?

$$
\mathbb{P}\left(\text{soccer}\right) = \frac{\text{soccer}}{\text{total}} = \frac{3}{10} = 0.3
$$

The numerator corresponds to the **event**, that's the size of the event. The denominatory corresponds to the size of the **sample space**, the total number of possible outcomes.

Here is a Venn diagram:

<img src='https://raw.githubusercontent.com/PaulToronto/DeepLearning-AI---Mathematics-for-Machine-Learning-and-Data-Science/refs/heads/main/images/venn.png'/>

#### Example

In probability, an **experiment** is an process that produces an outcome that is uncertain. A coin toss is an experiment. The coin is **fair**, meaning Heads and Tails are equally likely.

$$
\mathbb{P}\left(\text{heads}\right) = \frac{1}{2} = 0.5
$$

If we toss two coins, what is the probablity that both of them will be heads?

This experiment has 4 possible outcomes as shown here:

<img src='https://raw.githubusercontent.com/PaulToronto/DeepLearning-AI---Mathematics-for-Machine-Learning-and-Data-Science/refs/heads/main/images/four.png'/>

The probablity is calculated as follows:

$$
\mathbb{P}\left(\text{HH}\right) = \frac{1}{4} = 0.25
$$

What is $\mathbb{P}\left(\text{HHH}\right)$ when we have 3 coins?

We know there are $2^3 = 8$ possible outcomes, so we can construct the following table:



In [2]:
three_coins = pd.DataFrame({
    'coin_1': ['H', 'H', 'H', 'H', 'T', 'T', 'T', 'T'],
    'coin_2': ['H', 'H', 'T', 'T', 'H', 'H', 'T', 'T'],
    'coin_3': ['H', 'T', 'H', 'T', 'H', 'T', 'H', 'T']
})

three_coins

coin_1 coin_2 coin_3
0      H      H      H
1      H      H      T
2      H      T      H
3      H      T      T
4      T      H      H
5      T      H      T
6      T      T      H
7      T      T      T

In [3]:
event = three_coins[(three_coins['coin_1'] == 'H') &
                    (three_coins['coin_2'] == 'H') &
                    (three_coins['coin_3'] == 'H')]

event

coin_1 coin_2 coin_3
0      H      H      H

In [4]:
len(event) / len(three_coins)

0.125

$$
\mathbb{P}\left(\text{HHH}\right) = \frac{1}{8} = 0.125
$$

#### Dice Example

When you roll a fair 6-sided dice what is the probability of a 6?

$$
\mathbb{P}(6) = \frac{1}{6} = 0.1\dot{6}
$$

When you roll two dice, what is the probability of getting two 6s?

$$
\mathbb{P}(6, 6) = \frac{1}{36}
$$

### 3.1.1.2 Complement of Probability

#### Complement Rule

$$
\begin{align}
\mathbb{P}(A) &+ \mathbb{P}{(A^{\prime})} = 1 \\ \\
\mathbb{P}(A) &= 1 - \mathbb{P}{(A^{\prime})} \\
\mathbb{P}{(A^{\prime})} &= 1 - \mathbb{P}(A)
\end{align}
$$

#### Example

Whan flipping a coin three times, what is the probability of not obtaining 3 heads?

$$
\begin{align}
\mathbb{P}\left((HHH)^{\prime}\right) &= 1 - \mathbb{P}\left(HHH\right) \\
&= 1 - \frac{1}{8} \\
&= \frac{7}{8}
\end{align}
$$

#### Dice Example

You roll a dice. What is the probability of obtaining anything other than a 6?

$$
\begin{align}
\mathbb{P}(6^{\prime}) &= 1 - \mathbb{P}(6) \\
&= 1 - \frac{1}{6} \\
&= \frac{5}{6}
\end{align}
$$

### 3.1.1.3 Sum of Probabilities (Disjoint Events)

#### Example

If you roll a dice, what is the probablity of getting a 2 or a 3?

$$
\begin{align}
\mathbb{P}\left(2\cup3\right) &= \mathbb{P}\left(2\right) + \mathbb{P}\left(3\right) \\
&= \frac{1}{6} + \frac{1}{6} \\
&= \frac{1}{3}
\end{align}
$$

This addition only works if the events are disjoint.

#### Example

At a school, kids can **only play one sport**. The sport can be soccer or basketball. We are given the following:

$$
\begin{align}
\mathbb{P}\left(\text{soccer}\right) &= 0.3 \\
\mathbb{P}\left(\text{basketball}\right) &= 0.4
\end{align}
$$

What is the probability that a kid plays soccer or basketball?

$$
\begin{align}
\mathbb{P}\left(\text{soccer} \cup \text{basketball}\right) &= \mathbb{P}\left(\text{soccer}\right) + \mathbb{P}\left(\text{basketball}\right) \\
&= 0.3 + 0.4 \\
&= 0.7
\end{align}
$$

<img src='https://raw.githubusercontent.com/PaulToronto/DeepLearning-AI---Mathematics-for-Machine-Learning-and-Data-Science/refs/heads/main/images/prob1.png'/>

<img src='https://raw.githubusercontent.com/PaulToronto/DeepLearning-AI---Mathematics-for-Machine-Learning-and-Data-Science/refs/heads/main/images/prob2.png'/>



#### Dice Example

What is the probability of obtaining an even number or a 5?

$$
\begin{align}
\mathbb{P}\left(\text{even} \cup 5\right) &= \mathbb{P}\left(\text{even}\right) + \mathbb{P}\left(5\right)  \\
&= \frac{3}{6} + \frac{1}{6} \\
&= \frac{2}{3}
\end{align}
$$

This only works because a number **can't be both even and a 5**. These two events are disjoint.

#### Dice Example

Roll two dice. What is the probability of obtaining a sum of 7 or a sum of 10?

$$
\begin{align}
\mathbb{P}\left(\text{(sum 7)} \cup \text{(sum 10)}\right) &= \mathbb{P}\left(\text{sum 7}\right) + \mathbb{P}\left(\text{sum 10}\right) \\
&= \frac{6}{36} + \frac{3}{36} \\
&= \frac{9}{36} \\
&= \frac{1}{4}
\end{align}
$$

#### Dice Example

Roll two dice. What is the probability of obtaining a difference of 2 or a differnce of 1?

$$
\begin{align}
\mathbb{P}\left(\text{(diff 2)} \cup \text{(diff 1)}\right) &= \frac{8}{36} + \frac{10}{36} \\
&= \frac{18}{36} \\
&= \frac{1}{2}
\end{align}
$$

#### Sum Rule for Disjoint Events (Mutually exclusive)

$$
\begin{align}
\mathbb{P}\left(A \cup B\right) = \mathbb{P}\left(A\right) + \mathbb{P}\left(B\right)
\end{align}
$$

### 3.1.1.4 Sum of Probabilities (Joint Events)

The probability of rain is 0.8 and the probability of windy is 0.7.

$$
\begin{align}
\mathbb{P}\left(\text{rain}\right) &= 0.8 \\
\mathbb{P}\left(\text{wind}\right) &= 0.7
\end{align}
$$

What is $\mathbb{P}\left(\text{rain} \cup \text{wind}\right)$?

If we add them together as we did before, we get $1.5$, but that is greater than $1$ so that can't be right. It doesn't work because the events are **not disjoint**. It is possible to have both rain and wind at the same time.



#### Example


At a school, kids can **play as many sports as he wants**. We are given the following:

$$
\begin{align}
\mathbb{P}\left(\text{soccer}\right) &= 0.6 \\
\mathbb{P}\left(\text{basketball}\right) &= 0.5
\end{align}
$$

What is the probability that a kid plays soccer or basketball?

At this point, we don't have enough information to answer the question.

<img src='https://raw.githubusercontent.com/PaulToronto/DeepLearning-AI---Mathematics-for-Machine-Learning-and-Data-Science/refs/heads/main/images/soccer.png'/>

$$
\mathbb{P}\left(\text{soccer} \cup \text{basketball}\right) = \mathbb{P}\left(\text{soccer}\right) + \mathbb{P}\left(\text{basketball}\right) - \mathbb{P}\left(\text{soccer} \cap \text{basketball}\right)
$$

To solve this, we need to know the value of $\mathbb{P}\left(\text{soccer} \cap \text{basketball}\right) $

#### Example

6 kids play soccer, 5 kids play basketball and 3 kids play both soccer and basketball.

How many kids play soccer or basketball?

$$
6 + 5 - 3 = 8
$$

<img src='https://raw.githubusercontent.com/PaulToronto/DeepLearning-AI---Mathematics-for-Machine-Learning-and-Data-Science/refs/heads/main/images/venn2.png'/>

$$
\begin{align}
\left|S \cup B\right| &= \left|S\right| + \left|B\right| - \left|S \cap B\right| \\
&= 6 + 5 - 3 \\
&= 8
\end{align}
$$

This is also known as the **inclusion-exclusion principle**.

#### Example

At a school, kids can **play as many sports as he wants**. We are given the following:

$$
\begin{align}
\mathbb{P}\left(\text{soccer}\right) &= 0.6 \\
\mathbb{P}\left(\text{basketball}\right) &= 0.5 \\
\mathbb{P}\left(\text{soccer} \cap \text{basketball}\right) &= 0.3
\end{align}
$$

What is the probability that kid plays soccer or basketball?

$$
\begin{align}
\mathbb{P}\left(\text{soccer} \cup \text{basketball}\right) &= \mathbb{P}\left(\text{soccer}\right) + \mathbb{P}\left(\text{basketball}\right) - \mathbb{P}\left(\text{soccer} \cap \text{basketball}\right) \\
&= 0.6 + 0.5 - 0.3 \\
&= 0.8
\end{align}
$$


#### Sum Rule for Joint Events (Non-mutually exclusive)

$$
\begin{align}
\mathbb{P}\left(A \cup B\right) = \mathbb{P}\left(A\right) + \mathbb{P}\left(B\right) - \mathbb{P}\left(A \cap B\right)
\end{align}
$$

#### Dice Example

What is teh probability of obtaining a sum of 7 or a difference of 1?

$$
\begin{align}
\mathbb{P}\left(\text{sum7} \cup \text{diff1}\right) &= \mathbb{P}\left(\text{sum7}\right) + \mathbb{P}\left(\text{diff1}\right) - \mathbb{P}\left(\text{sum7} \cap \text{diff1}\right) \\
&= \frac{6}{36} + \frac{10}{36} - \frac{2}{36} \\
&= \frac{14}{36} \\
&= \frac{7}{18}
\end{align}
$$

### 3.1.1.5 Independence

If I toss a coin twice, the outcome of the first toss has no effect on the outcome of the second toss. On the other hand, if I'm playing chess, what happened on the $10^{\text{th}}$ move does affect what happens on the $11^{\text{th}}$ move.

Understanding **independence** helps us simplify calculations.

#### Example

A school has 100 kids. 50 of them like playing soccer and the other 50 don't like playing soccer. The kids are randomly split into two equal groups of size 50. Based on this, what is your estimate of the number of kids in group 1 who would like soccer?

My answer: 25

#### Example

In a school of 100 kids, 40 of them like soccer and 60 don't. Again, the kids are randomly split into two groups. Group 1 has 30 kids and Group 2 has 70.

What is your best estimate about how many kids in Group 1 like to play soccer?

My answer: $0.4 \times 30 = 12$

$$
\begin{align}
\mathbb{P}\left(\text{soccer} \cap \text{group1}\right) &= \mathbb{P}\left(\text{soccer}\right) \cdot \mathbb{P}\left(\text{group1}\right) \\
&= 0.4 \cdot 0.3 \\
&= 0.12
\end{align}
$$

#### Product Rule (for Independent Events)

$$
\mathbb{P}\left(A \cap B\right) = \mathbb{P}\left(A\right) \cdot \mathbb{P}\left(B\right)
$$

#### Coin Example

Consider tossing a fair coin 5 times. What is the probability that the coin lands on heads all 5 times?

$$
\begin{align}
\mathbb{P}\left(\text{5heads}\right) &= \left(\frac{1}{2}\right)^{5} \\
&= \frac{1}{32}
\end{align}
$$

#### Dice Example

When two dice are rolled, what is the probability of obtaining two sixes?

$$
\begin{align}
\mathbb{P}\left(\text{2_sixes}\right) &= \frac{1}{6} \cdot \frac{1}{6} \\
&= \frac{1}{36}
\end{align}
$$

#### Dice Example

If you roll a fair die 10 times, what is the probabilty of getting 10 sixes?

$$
\begin{align}
\mathbb{P}\left(\text{10_sixes}\right) &= \left(\frac{1}{6}\right)^{10} \\
&= \frac{1}{60466176}
\end{align}
$$

### 3.1.1.6 Birthday Problem